sentinel-core/src/main/java/com/alibaba/csp/sentinel/spi/SpiLoader.java

```diff
@@ -0,0 +1,541 @@
+                    }
+                });
+            } catch (IOException e) {
+                fail("error reading SPI file", e);
+            } finally {
+                closeResources(in, br);
+            }
+        }
+    }
+
+    @Override
+    public String toString() {
+        return "com.alibaba.csp.sentinel.spi.SpiLoader[" + service.getName() + "]";
+    }
+
+    /**
+     * Create Provider instance list
+     *
+     * @param clazzList class types of Providers
+     * @return Provider instance list
+     */
+    private List<S> createInstanceList(List<Class<? extends S>> clazzList) {
+        if (clazzList == null || clazzList.size() == 0) {
+            return Collections.emptyList();
+        }
+
+        List<S> instances = new ArrayList<>(clazzList.size());
+        for (Class<? extends S> clazz : clazzList) {
+            S instance = createInstance(clazz);
+            instances.add(instance);
+        }
+        return instances;
+    }
+
+    /**
+     * Create Provider instance
+     *
+     * @param clazz class type of Provider
+     * @return Provider class
+     */
+    private S createInstance(Class<? extends S> clazz) {
+        Spi spi = clazz.getAnnotation(Spi.class);
+        boolean singleton = true;
+        if (spi != null) {
+            singleton = spi.isSingleton();
+        }
+        return createInstance(clazz, singleton);
+    }
+
+    /**
+     * Create Provider instance
+     *
+     * @param clazz     class type of Provider
+     * @param singleton if instance is singleton or prototype
+     * @return Provider instance
+     */
+    private S createInstance(Class<? extends S> clazz, boolean singleton) {
+        S instance = null;
+        try {
+            if (singleton) {
+                instance = singletonMap.get(clazz.getName());
+                if (instance == null) {
+                    synchronized (this) {
+                        instance = service.cast(clazz.newInstance());
+                        singletonMap.put(clazz.getName(), instance);
+                    }
+                }
+            } else {
+                instance = service.cast(clazz.newInstance());
+            }
+        } catch (Throwable e) {
+            fail(clazz.getName() + " could not be instantiated");
+        }
+        return instance;
+    }
+
+    /**
+     * Close all resources
+     *
+     * @param closeables {@link Closeable} resources
+     */
+    private void closeResources(Closeable... closeables) {
+        if (closeables == null) {
+            return;
+        }
+
+        for (Closeable closeable : closeables) {
+            try {
+                closeable.close();
+            } catch (Exception e) {
+                fail("error closing SPI file", e);
+            }
+        }
+    }
+
+    /**
+     * Throw {@link SpiLoaderException} with message
+     *
+     * @param msg error message
+     */
+    private void fail(String msg) {
```

Comment:
I think `fail` is concise, while `logAndRaise` is more accurately.
The naming way is copy from `ServiceLoader#fail()`, besides throw Exception, add a `RecordLog.error` operation.

Reply:
> I think `fail` is concise, while `logAndRaise` is more accurately.
> The naming way is copy from `ServiceLoader#fail()`, besides throw Exception, add a `RecordLog.error` operation.

Because `fail` will throw `unchecked` exception which may lead actual workflow unexpected.
Whatever just a suggestion it's not required request of change. So you can make the final decision about it.